File: cross_language/csharp/Feed.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CrossLanguage
{
    public class FeedReader
    {
        // Documented return contract: Recent(n) returns an IEnumerable
        // that may be enumerated more than once and yields the same
        // items on each pass. Caches and audit replays depend on it.
        protected readonly List<string> Items;

        public FeedReader(IEnumerable<string> items)
        {
            Items = items.ToList();
        }

        public virtual IEnumerable<string> Recent(int n)
        {
            int from = Math.Max(0, Items.Count - n);
            return Items.GetRange(from, Items.Count - from);
        }
    }

    public class StreamingFeedReader : FeedReader
    {
        // Covariant return abuse via C# 9 covariant return types: the
        // override narrows IEnumerable<string> to a single-pass iterator
        // backed by `yield return`. The compiler accepts it because
        // IEnumerator<T> is structurally compatible at the call site of
        // covariant overrides, but callers that walked the parent's
        // return value twice silently observe an empty second pass.
        public StreamingFeedReader(IEnumerable<string> items) : base(items) { }

        public override IEnumerable<string> Recent(int n)
        {
            int from = Math.Max(0, Items.Count - n);
            for (int i = from; i < Items.Count; i++)
            {
                yield return Items[i];
            }
        }
    }

    public static class FeedDriver
    {
        public static (int First, int Second) SummarizeTwice(FeedReader reader, int n)
        {
            // Encodes the documented re-enumerable contract: ask for a
            // window, walk it twice, expect matching counts. Against a
            // streaming subclass the second walk collapses to zero
            // because the underlying iterator is exhausted.
            IEnumerable<string> window = reader.Recent(n);
            int first = window.Count();
            int second = window.Count();
            return (first, second);
        }

        public static void Run()
        {
            var seed = new[] { "a", "b", "c" };
            var baseResult = SummarizeTwice(new FeedReader(seed), 3);
            var streamingResult = SummarizeTwice(new StreamingFeedReader(seed), 3);
            Console.WriteLine($"base first={baseResult.First} second={baseResult.Second}");
            Console.WriteLine($"streaming first={streamingResult.First} second={streamingResult.Second}");
        }
    }
}

File: src/notifications/feed.py
from __future__ import annotations

from typing import Iterable, Iterator


class FeedReader:
    # Documented return contract: recent(n) returns a re-iterable
    # sequence. Callers may walk it more than once and observe the same
    # items each pass — caches and audit replays rely on this.
    def __init__(self, items: list[str]) -> None:
        self._items = list(items)

    def recent(self, n: int) -> Iterable[str]:
        return list(self._items[-n:])


class StreamingFeedReader(FeedReader):
    # Covariant return abuse: Iterator[str] is a structural subtype of
    # Iterable[str], so the type checker accepts the override. But an
    # iterator is single-pass — callers that walk the return value
    # twice silently observe an empty sequence on the second walk.
    def recent(self, n: int) -> Iterator[str]:
        for item in self._items[-n:]:
            yield item


def summarize_twice(reader: FeedReader, n: int) -> tuple[int, int]:
    # Encodes the documented re-iterable contract: the caller asks for a
    # window once, then walks the result twice. Against the base both
    # counts must match.
    items = reader.recent(n)
    first = sum(1 for _ in items)
    second = sum(1 for _ in items)
    return first, second
